From: Brian May <bam@debian.org>
Date: Thu, 21 Oct 2021 10:05:16 +1100
Subject: kdc: validate sname in TGS-REQ

Fixes CVE-2021-3671

In tgs_build_reply(), validate the server name in the TGS-REQ is present before
dereferencing.

Sources:

* https://github.com/heimdal/heimdal/commit/04171147948d0a3636bc6374181926f0fb2ec83a
* https://github.com/heimdal/heimdal/commit/773802aecfb4b6a73817fa522faeb55b2a7cdb2a
---
 kdc/krb5tgs.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/kdc/krb5tgs.c b/kdc/krb5tgs.c
index 89bd1f2..0ad9e6c 100644
--- a/kdc/krb5tgs.c
+++ b/kdc/krb5tgs.c
@@ -1660,6 +1660,10 @@ tgs_build_reply(krb5_context context,
 
 	s = &adtkt.cname;
 	r = adtkt.crealm;
+    } else if (s == NULL) {
+	ret = KRB5KDC_ERR_S_PRINCIPAL_UNKNOWN;
+	krb5_set_error_message(context, ret, "No server in request");
+	goto out;
     }
 
     _krb5_principalname2krb5_principal(context, &sp, *s, r);
